// @sanity/block-tools/lib/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_SUPPORTED_ANNOTATIONS = exports.DEFAULT_SUPPORTED_DECORATORS = exports.DEFAULT_SUPPORTED_STYLES = exports.ELEMENT_MAP = exports.HTML_LIST_ITEM_TAGS = exports.HTML_DECORATOR_TAGS = exports.HTML_MISC_TAGS = exports.HTML_HEADER_TAGS = exports.HTML_LIST_CONTAINER_TAGS = exports.HTML_SPAN_TAGS = exports.HTML_BLOCK_TAGS = exports.DEFAULT_SPAN = exports.DEFAULT_BLOCK = exports.SLATE_DEFAULT_BLOCK = exports.BLOCK_DEFAULT_STYLE = undefined;

var _uniq2 = require('lodash/uniq');

var _uniq3 = _interopRequireDefault(_uniq2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BLOCK_DEFAULT_STYLE = exports.BLOCK_DEFAULT_STYLE = 'normal';

// Slate block
const SLATE_DEFAULT_BLOCK = exports.SLATE_DEFAULT_BLOCK = Object.freeze({
  kind: 'block',
  type: 'contentBlock',
  data: {
    style: BLOCK_DEFAULT_STYLE
  }
});

const DEFAULT_BLOCK = exports.DEFAULT_BLOCK = Object.freeze({
  _type: 'block',
  markDefs: [],
  style: BLOCK_DEFAULT_STYLE
});

const DEFAULT_SPAN = exports.DEFAULT_SPAN = Object.freeze({
  _type: 'span',
  marks: []
});

const HTML_BLOCK_TAGS = exports.HTML_BLOCK_TAGS = {
  p: DEFAULT_BLOCK,
  blockquote: _extends({}, DEFAULT_BLOCK, { style: 'blockquote' })
};

const HTML_SPAN_TAGS = exports.HTML_SPAN_TAGS = {
  span: { kind: 'text' }
};

const HTML_LIST_CONTAINER_TAGS = exports.HTML_LIST_CONTAINER_TAGS = {
  ol: { kind: null },
  ul: { kind: null }
};

const HTML_HEADER_TAGS = exports.HTML_HEADER_TAGS = {
  h1: _extends({}, DEFAULT_BLOCK, { style: 'h1' }),
  h2: _extends({}, DEFAULT_BLOCK, { style: 'h2' }),
  h3: _extends({}, DEFAULT_BLOCK, { style: 'h3' }),
  h4: _extends({}, DEFAULT_BLOCK, { style: 'h4' }),
  h5: _extends({}, DEFAULT_BLOCK, { style: 'h5' }),
  h6: _extends({}, DEFAULT_BLOCK, { style: 'h6' })
};

const HTML_MISC_TAGS = exports.HTML_MISC_TAGS = {
  br: _extends({}, DEFAULT_BLOCK, { style: 'normal' })
};
const HTML_DECORATOR_TAGS = exports.HTML_DECORATOR_TAGS = {
  b: 'strong',
  strong: 'strong',

  i: 'em',
  em: 'em',

  u: 'underline',
  s: 'strike-through',
  strike: 'strike-through',
  del: 'strike-through',

  code: 'code'
};

const HTML_LIST_ITEM_TAGS = exports.HTML_LIST_ITEM_TAGS = {
  li: _extends({}, DEFAULT_BLOCK, {
    style: 'normal',
    level: 1,
    listItem: 'bullet'
  })
};

const ELEMENT_MAP = exports.ELEMENT_MAP = _extends({}, HTML_BLOCK_TAGS, HTML_SPAN_TAGS, HTML_LIST_CONTAINER_TAGS, HTML_LIST_ITEM_TAGS, HTML_HEADER_TAGS, HTML_MISC_TAGS);

const DEFAULT_SUPPORTED_STYLES = exports.DEFAULT_SUPPORTED_STYLES = (0, _uniq3.default)(Object.keys(ELEMENT_MAP).filter(tag => ELEMENT_MAP[tag].style).map(tag => ELEMENT_MAP[tag].style));

const DEFAULT_SUPPORTED_DECORATORS = exports.DEFAULT_SUPPORTED_DECORATORS = (0, _uniq3.default)(Object.keys(HTML_DECORATOR_TAGS).map(tag => HTML_DECORATOR_TAGS[tag]));

const DEFAULT_SUPPORTED_ANNOTATIONS = exports.DEFAULT_SUPPORTED_ANNOTATIONS = ['link'];
