// @sanity/preview/lib/components/PreviewFields.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreviewFields;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _PreviewSubscriber = require('./PreviewSubscriber');

var _PreviewSubscriber2 = _interopRequireDefault(_PreviewSubscriber);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function arrify(val) {
  if (Array.isArray(val)) {
    return val;
  }
  return typeof val === undefined ? [] : [val];
}

function PreviewFields(props) {
  return _react2.default.createElement(
    _PreviewSubscriber2.default,
    { value: props.document, type: props.type, fields: arrify(props.fields) },
    ({ snapshot }) => _react2.default.createElement(
      'span',
      null,
      snapshot && props.children(snapshot)
    )
  );
}

PreviewFields.propTypes = {
  document: _propTypes2.default.object,
  fields: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  type: _propTypes2.default.object.isRequired,
  children: _propTypes2.default.func
};
