// @sanity/form-builder/lib/utils/withDocument.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = withDocument;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function withDocument(ComposedComponent /*: any*/) {
  var _class, _temp;

  return _temp = _class = class WithDocument extends _react2.default.PureComponent {

    constructor(props /*: any*/, context /*: any*/) {
      super();

      this.setInput = input => {
        this._input = input;
      };

      const formBuilder = context.formBuilder;

      this.state = { document: formBuilder.getDocument() };
      this.unsubscribe = formBuilder.onPatch(({ snapshot }) => {
        // we will also receive "delete"-patches, with {snapshot: null}. Don't pass null documents.
        if (snapshot) {
          this.setState({ document: snapshot });
        }
      });
    }
    componentWillUnmount() {
      this.unsubscribe();
    }
    focus() {
      this._input.focus();
    }

    render() {
      return _react2.default.createElement(ComposedComponent, _extends({ ref: this.setInput, document: this.state.document }, this.props));
    }
  }, _class.displayName = `withDocument(${ComposedComponent.displayName || ComposedComponent.name})`, _class.contextTypes = {
    formBuilder: _propTypes2.default.any
  }, _temp;
}
