// @sanity/form-builder/lib/utils/withValuePath.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = withValuePath;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function withValuePath(ComposedComponent /*: any*/) {
  var _class, _temp2;

  return _temp2 = _class = class WithValuePath extends _react2.default.PureComponent {
    constructor(...args) {
      var _temp;

      return _temp = super(...args), this.setInput = input => {
        this._input = input;
      }, _temp;
    }

    focus() {
      this._input.focus();
    }

    render() {
      const getValuePath = this.context.getValuePath;

      return _react2.default.createElement(ComposedComponent, _extends({ ref: this.setInput }, this.props, { getValuePath: getValuePath }));
    }
  }, _class.displayName = `withValuePath(${ComposedComponent.displayName || ComposedComponent.name})`, _class.contextTypes = {
    getValuePath: _propTypes2.default.func,
    formBuilder: _propTypes2.default.any
  }, _temp2;
}
