// @sanity/form-builder/lib/sanity/inputResolver/defaultInputs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _BooleanInput = require('../../inputs/BooleanInput');

var _BooleanInput2 = _interopRequireDefault(_BooleanInput);

var _EmailInput = require('../../inputs/EmailInput');

var _EmailInput2 = _interopRequireDefault(_EmailInput);

var _NumberInput = require('../../inputs/NumberInput');

var _NumberInput2 = _interopRequireDefault(_NumberInput);

var _ObjectInput = require('../../inputs/ObjectInput');

var _ObjectInput2 = _interopRequireDefault(_ObjectInput);

var _StringInput = require('../../inputs/StringInput');

var _StringInput2 = _interopRequireDefault(_StringInput);

var _DateTimeInput = require('../../inputs/DateTimeInput');

var _DateTimeInput2 = _interopRequireDefault(_DateTimeInput);

var _TextInput = require('../../inputs/TextInput');

var _TextInput2 = _interopRequireDefault(_TextInput);

var _UrlInput = require('../../inputs/UrlInput');

var _UrlInput2 = _interopRequireDefault(_UrlInput);

var _SlugInput = require('../../inputs/Slug/SlugInput');

var _SlugInput2 = _interopRequireDefault(_SlugInput);

var _SanityArrayInput = require('../inputs/SanityArrayInput');

var _SanityArrayInput2 = _interopRequireDefault(_SanityArrayInput);

var _SanityImageInput = require('../inputs/SanityImageInput');

var _SanityImageInput2 = _interopRequireDefault(_SanityImageInput);

var _SanityFileInput = require('../inputs/SanityFileInput');

var _SanityFileInput2 = _interopRequireDefault(_SanityFileInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  object: _ObjectInput2.default,
  array: _SanityArrayInput2.default,
  boolean: _BooleanInput2.default,
  number: _NumberInput2.default,
  text: _TextInput2.default,
  email: _EmailInput2.default,
  datetime: _DateTimeInput2.default,
  url: _UrlInput2.default,
  image: _SanityImageInput2.default,
  file: _SanityFileInput2.default,
  string: _StringInput2.default,
  slug: _SlugInput2.default
};
