// @sanity/form-builder/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _FormBuilder = require('./FormBuilder');

Object.defineProperty(exports, 'FormBuilder', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormBuilder).default;
  }
});

var _FormBuilderContext = require('./FormBuilderContext');

Object.defineProperty(exports, 'FormBuilderContext', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormBuilderContext).default;
  }
});

var _BlockEditorSlate = require('./inputs/BlockEditor-slate');

Object.defineProperty(exports, 'BlockEditor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_BlockEditorSlate).default;
  }
});

var _ReferenceInput = require('./inputs/ReferenceInput');

Object.defineProperty(exports, 'ReferenceInput', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ReferenceInput).default;
  }
});

var _FileInput = require('./inputs/FileInput');

Object.defineProperty(exports, 'FileInput', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FileInput).default;
  }
});

var _ImageInput = require('./inputs/ImageInput');

Object.defineProperty(exports, 'ImageInput', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ImageInput).default;
  }
});
exports.createFormBuilder = createFormBuilder;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createFormBuilder() {
  throw new Error('The factory function createFormBuilder(...) has been removed. Please use <FormBuilder .../> instead');
}
