// @sanity/schema/lib/legacy/types/boolean.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BooleanType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _primitivePreview = require('../preview/primitivePreview');

var _primitivePreview2 = _interopRequireDefault(_primitivePreview);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options'];

const BOOLEAN_CORE = {
  name: 'boolean',
  type: null,
  jsonType: 'boolean'
};

const BooleanType = exports.BooleanType = {
  get() {
    return BOOLEAN_CORE;
  },
  extend(subTypeDef) {
    const parsed = Object.assign((0, _pick3.default)(BOOLEAN_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: BOOLEAN_CORE,
      preview: _primitivePreview2.default
    });

    return subtype(parsed);

    function subtype(parent) {
      return {
        get() {
          return parent;
        },
        extend: extensionDef => {
          const current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), {
            type: parent
          });
          return subtype(current);
        }
      };
    }
  }
};
