// @sanity/schema/lib/legacy/types/image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageType = undefined;

var _startCase2 = require('lodash/startCase');

var _startCase3 = _interopRequireDefault(_startCase2);

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _utils = require('./utils');

var _fieldDefs = require('./image/fieldDefs');

var _createPreviewGetter = require('../preview/createPreviewGetter');

var _createPreviewGetter2 = _interopRequireDefault(_createPreviewGetter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

const OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options', 'fieldsets'];

const IMAGE_CORE = {
  name: 'image',
  type: null,
  jsonType: 'object'
};

const DEFAULT_OPTIONS = {};

const ImageType = {
  get() {
    return IMAGE_CORE;
  },
  extend(subTypeDef, extendMember) {
    const options = _extends({}, subTypeDef.options || DEFAULT_OPTIONS);

    const fields = (subTypeDef.fields || []).concat([options.hotspot && _fieldDefs.HOTSPOT_FIELD, options.hotspot && _fieldDefs.CROP_FIELD, _fieldDefs.ASSET_FIELD]).filter(Boolean);

    const parsed = Object.assign((0, _pick3.default)(IMAGE_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: IMAGE_CORE,
      title: subTypeDef.title || (0, _startCase3.default)(subTypeDef.name || subTypeDef.type || ''),
      options: options
    });

    (0, _utils.lazyGetter)(parsed, 'fields', () => {
      return fields.map(fieldDef => {
        const name = fieldDef.name,
              type = _objectWithoutProperties(fieldDef, ['name']);

        return {
          name: name,
          type: extendMember(type)
        };
      });
    });

    (0, _utils.lazyGetter)(parsed, 'preview', (0, _createPreviewGetter2.default)(Object.assign({}, subTypeDef, { fields })));

    return subtype(parsed);

    function subtype(parent) {
      return {
        get() {
          return parent;
        },
        extend: extensionDef => {
          if (extensionDef.fields) {
            throw new Error('Cannot override `fields` of subtypes of "image"');
          }
          const current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), {
            type: parent
          });
          return subtype(current);
        }
      };
    }
  }
};
exports.ImageType = ImageType;
