// @sanity/core/lib/commands/dataset/datasetVisibilityCommand.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

exports.default = {
  name: 'visibility',
  group: 'dataset',
  signature: 'get/set [dataset] [mode]',
  description: 'Set visibility of a dataset',
  // eslint-disable-next-line complexity
  action: (() => {
    var _ref = _asyncToGenerator(function* (args, context) {
      const apiClient = context.apiClient,
            output = context.output;

      var _args$argsWithoutOpti = _slicedToArray(args.argsWithoutOptions, 3);

      const action = _args$argsWithoutOpti[0],
            ds = _args$argsWithoutOpti[1],
            aclMode = _args$argsWithoutOpti[2];

      const client = apiClient();

      if (!client.datasets.edit) {
        throw new Error('@sanity/cli must be upgraded first:\n  npm install -g @sanity/cli');
      }

      if (!action) {
        throw new Error('Action must be provided (get/set)');
      }

      if (!['set', 'get'].includes(action)) {
        throw new Error('Invalid action (only get/set allowed)');
      }

      if (!ds) {
        throw new Error('Dataset name must be provided');
      }

      if (action === 'set' && !aclMode) {
        throw new Error('Please provide a visibility mode (public/private)');
      }

      const dataset = `${ds}`;
      const current = (yield client.datasets.list()).find(function (curr) {
        return curr.name === dataset;
      });

      if (!current) {
        throw new Error('Dataset not found');
      }

      if (action === 'get') {
        output.print(current.aclMode);
        return;
      }

      if (current.aclMode === aclMode) {
        output.print(`Dataset already in "${aclMode}"-mode`);
        return;
      }

      if (aclMode === 'private') {
        output.print('Please note that while documents are private, assets (files and images) are still public\n');
      }

      yield client.datasets.edit(dataset, { aclMode });
      output.print('Dataset visibility changed');
    });

    return function action(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })()
};
