// @sanity/form-builder/lib/sanity/inputs/SanityReferenceInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reference = require('./client-adapters/reference');

var _ReferenceInput = require('../../inputs/ReferenceInput');

var _ReferenceInput2 = _interopRequireDefault(_ReferenceInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SanityReference extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.setInput = input => {
      this._input = input;
    }, _temp;
  }

  focus() {
    this._input.focus();
  }
  render() {
    return _react2.default.createElement(_ReferenceInput2.default, _extends({}, this.props, {
      onSearch: _reference.search,
      getPreviewSnapshot: _reference.getPreviewSnapshot,
      ref: this.setInput
    }));
  }
}
exports.default = SanityReference;
