// @sanity/form-builder/lib/sanity/inputs/SanityImageInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ = require('../..');

var _assets = require('./client-adapters/assets');

var _resolveUploader = require('../uploads/resolveUploader');

var _resolveUploader2 = _interopRequireDefault(_resolveUploader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SanityImageInput extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.setInput = input => {
      this._input = input;
    }, _temp;
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    return _react2.default.createElement(_.ImageInput, _extends({}, this.props, {
      resolveUploader: _resolveUploader2.default,
      materialize: _assets.materializeReference,
      ref: this.setInput
    }));
  }
}
exports.default = SanityImageInput;
