// @sanity/components/lib/utilities/Portal/LegacyPortal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LegacyPortal = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LegacyPortal extends _react2.default.Component {
  componentDidMount() {
    this.renderPortal();
  }

  componentDidUpdate(props) {
    this.renderPortal();
  }

  componentWillUnmount() {
    if (this.node) {
      // Unmount the children rendered in custom node
      _reactDom2.default.unmountComponentAtNode(this.node);
      document.body.removeChild(this.node);
      this.node = null;
      this.portal = null;
    }
  }

  renderPortal() {
    if (!this.node) {
      this.node = document.createElement('div');
      document.body.appendChild(this.node);
    }

    let children = this.props.children;
    // https://gist.github.com/jimfb/d99e0678e9da715ccf6454961ef04d1b
    if (typeof this.props.children.type === 'function') {
      children = _react2.default.cloneElement(this.props.children);
    }

    this.portal = _reactDom2.default.unstable_renderSubtreeIntoContainer(this, children, this.node);
  }

  render() {
    return null;
  }
}
exports.LegacyPortal = LegacyPortal; // This file is a fallback for a consumer who is not yet on React 16
// as createPortal was introduced in React 16

// based on tajo/react-portal, but removes unused functionality and includes a bugfix

LegacyPortal.propTypes = {
  children: _propTypes2.default.node.isRequired
};
