// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/AnnotationButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _button = require('part:@sanity/components/toggles/button');

var _button2 = _interopRequireDefault(_button);

var _linkIcon = require('part:@sanity/base/link-icon');

var _linkIcon2 = _interopRequireDefault(_linkIcon);

var _AnnotationButton = require('./styles/AnnotationButton.css');

var _AnnotationButton2 = _interopRequireDefault(_AnnotationButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class AnnotationButton extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.handleToggleButtonClick = () => {
      this.props.onClick(this.props.annotation);
    }, _temp;
  }

  render() {
    const annotation = this.props.annotation;

    return _react2.default.createElement(_button2.default, {
      onClick: this.handleToggleButtonClick,
      title: annotation.type.title,
      disabled: annotation.disabled,
      selected: annotation.active,
      className: _AnnotationButton2.default.button,
      icon: _linkIcon2.default
    });
  }
}
exports.default = AnnotationButton;
AnnotationButton.propTypes = {
  onClick: _propTypes2.default.func,
  annotation: _propTypes2.default.object
};
