// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/BlockStyle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.blockStyleShape = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _style = require('part:@sanity/components/selects/style');

var _style2 = _interopRequireDefault(_style);

var _BlockStyle = require('./styles/BlockStyle.css');

var _BlockStyle2 = _interopRequireDefault(_BlockStyle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const blockStyleShape = exports.blockStyleShape = _propTypes2.default.shape({
  key: _propTypes2.default.string,
  active: _propTypes2.default.bool,
  title: _propTypes2.default.string,
  preview: _propTypes2.default.node
});

class BlockStyle extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.renderItem = item => {
      return item.preview;
    }, _temp;
  }

  render() {
    if (!this.props.items || this.props.items.length === 0) {
      return null;
    }
    return _react2.default.createElement(
      'label',
      { className: _BlockStyle2.default.root },
      _react2.default.createElement(
        'span',
        { style: { display: 'none' } },
        'Text'
      ),
      _react2.default.createElement(_style2.default, {
        items: this.props.items,
        value: this.props.value,
        onChange: this.props.onSelect,
        renderItem: this.renderItem,
        transparent: true
      })
    );
  }
}
exports.default = BlockStyle;
BlockStyle.propTypes = {
  value: _propTypes2.default.arrayOf(blockStyleShape),
  items: _propTypes2.default.arrayOf(blockStyleShape),
  onSelect: _propTypes2.default.func
};
