// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/InsertBlocks.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.insertBlockShape = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _dropdown = require('part:@sanity/components/buttons/dropdown');

var _dropdown2 = _interopRequireDefault(_dropdown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const insertBlockShape = exports.insertBlockShape = _propTypes2.default.shape({
  type: _propTypes2.default.object,
  title: _propTypes2.default.string
});

class InsertBlocks extends _react2.default.Component {

  shouldComponentUpdate(nextProps, nextState) {
    return this.props.blocks !== nextProps.blocks;
  }

  render() {
    return _react2.default.createElement(
      _dropdown2.default,
      {
        items: this.props.blocks,
        onAction: this.props.onInsertBlock,
        kind: 'simple',
        origin: 'right'
      },
      'Insert'
    );
  }
}
exports.default = InsertBlocks;
InsertBlocks.propTypes = {
  onInsertBlock: _propTypes2.default.func,
  blocks: _propTypes2.default.arrayOf(insertBlockShape)
};
