// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/ListItems.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listItem = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _formatListBulletedIcon = require('part:@sanity/base/format-list-bulleted-icon');

var _formatListBulletedIcon2 = _interopRequireDefault(_formatListBulletedIcon);

var _formatListNumberedIcon = require('part:@sanity/base/format-list-numbered-icon');

var _formatListNumberedIcon2 = _interopRequireDefault(_formatListNumberedIcon);

var _sanityLogoIcon = require('part:@sanity/base/sanity-logo-icon');

var _sanityLogoIcon2 = _interopRequireDefault(_sanityLogoIcon);

var _button = require('part:@sanity/components/toggles/button');

var _button2 = _interopRequireDefault(_button);

var _ListItems = require('./styles/ListItems.css');

var _ListItems2 = _interopRequireDefault(_ListItems);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const listItem = exports.listItem = _propTypes2.default.shape({
  active: _propTypes2.default.bool,
  type: _propTypes2.default.string,
  title: _propTypes2.default.string
});

class ListItem extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.renderButton = listFormat => {
      const onClick = event => {
        this.props.onClick(listFormat.type, listFormat.active);
      };
      const Icon = this.getIcon(listFormat.type);
      return _react2.default.createElement(_button2.default, {
        key: `listButton${listFormat.type}`,
        selected: listFormat.active,
        onClick: onClick,
        title: listFormat.title,
        icon: Icon
      });
    }, _temp;
  }

  getIcon(type) {
    switch (type) {
      case 'number':
        return _formatListNumberedIcon2.default;
      case 'bullet':
        return _formatListBulletedIcon2.default;
      default:
        return _sanityLogoIcon2.default;
    }
  }

  render() {
    return _react2.default.createElement(
      'div',
      { className: _ListItems2.default.root },
      this.props.listItems.map(this.renderButton)
    );
  }
}
exports.default = ListItem;
ListItem.propTypes = {
  onClick: _propTypes2.default.func,
  listItems: _propTypes2.default.arrayOf(listItem)
};
