// @sanity/form-builder/lib/inputs/BlockEditor-slate/ItemForm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FormBuilderPropTypes = require('../../FormBuilderPropTypes');

var _FormBuilderPropTypes2 = _interopRequireDefault(_FormBuilderPropTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ItemForm extends _react2.default.PureComponent {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.handleChange = event => {
      const onChange = this.props.onChange;

      onChange(event);
    }, _temp;
  }

  resolveInputComponent(type, fieldType) {
    return this.context.formBuilder.resolveInputComponent(type, fieldType);
  }

  render() {
    var _props = this.props;
    const value = _props.value,
          type = _props.type,
          level = _props.level;


    const InputComponent = this.context.formBuilder.resolveInputComponent(type);
    if (!InputComponent) {
      return _react2.default.createElement(
        'div',
        null,
        'No input component resolved for type ',
        `"${type.name}"`
      );
    }

    return _react2.default.createElement(InputComponent, { value: value, type: type, level: level, onChange: this.handleChange });
  }
}
exports.default = ItemForm;
ItemForm.propTypes = {
  type: _FormBuilderPropTypes2.default.type.isRequired,
  value: _propTypes2.default.any,
  level: _propTypes2.default.number,
  onChange: _propTypes2.default.func
};
ItemForm.contextTypes = {
  formBuilder: _propTypes2.default.object
};
